pkg/scheduler/plugins/tdm/tdm.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	return
+}
+
+func (bp *tdmPlugin) availableRevocableZone(rz string) error {
+	// rzRaw format 00:00-23:59
+	rzRaw, ok := bp.args[rz]
+	if !ok {
+		return fmt.Errorf("revocable zone %v not support", rz)
+	}
+
+	now := time.Now()
+
+	start, end, err := parseRevocableZone(rzRaw)
+	if err != nil {
+		return err
+	}
+
+	if now.Unix() < start.Unix() || now.Unix() > end.Unix() {
+		return fmt.Errorf("current time beyond revocable zone %v:%v", rz, rzRaw)
+	}
+
+	return nil
+}
+
+func (bp *tdmPlugin) OnSessionOpen(ssn *framework.Session) {
+	klog.V(4).Infof("Enter tdm plugin ...")
+	if klog.V(4) {
+		defer func() {
+			klog.V(4).Infof("Leaving tdm plugin.")
+		}()
+	}
+
+	// tdm plugin just handle revocable node
+	predicateFn := func(task *api.TaskInfo, node *api.NodeInfo) error {
+		rz, ok := node.Node.Labels[RevocableZoneKey]
+		if !ok {
+			return nil
+		}
+
+		if err := bp.availableRevocableZone(rz); err != nil {
+			return fmt.Errorf("plugin %s predicates %w", bp.Name(), err)
+		}
+
+		if _, ok := task.Pod.Annotations[PreemptableTaskKey]; !ok {
+			msg := fmt.Sprintf("task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return fmt.Errorf("plugin %s predicates %s", bp.Name(), msg)
+		}
+
+		klog.V(4).Infof("TDM filter for Task %s/%s on node %s pass.", task.Namespace, task.Name, node.Name)
+		return nil
+	}
+
+	// tdm plugin just handle revocable node
+	nodeOrderFn := func(task *api.TaskInfo, node *api.NodeInfo) (float64, error) {
+		score := 0.0
+
+		rz, ok := node.Node.Labels[RevocableZoneKey]
+		if !ok {
+			return score, nil
+		}
+
+		if err := bp.availableRevocableZone(rz); err != nil {
+			klog.V(4).Infof("TDM not available %s", err)
+			return score, err
+		}
+
+		if _, ok := task.Pod.Annotations[PreemptableTaskKey]; !ok {
+			klog.V(4).Infof("TDM task %s/%s is not preemptable task, skip to schedule to node %s", task.Namespace, task.Name, node.Name)
+			return score, nil
+		}
+
+		score = float64(v1alpha1.MaxNodeScore)
+
+		klog.V(4).Infof("TDM score for Task %s/%s on node %s is: %v", task.Namespace, task.Name, node.Name, score)
+		return score, nil
+	}
+
+	preemptableFn := func(preemptor *api.TaskInfo, preemptees []*api.TaskInfo) []*api.TaskInfo {
+		if _, ok := preemptor.Pod.Annotations[PreemptableTaskKey]; ok {
+			klog.V(4).Infof("task %s/%s is preemptable, do nothing skip", preemptor.Namespace, preemptor.Name)
+			return preemptees
+		}
+
+		var victims []*api.TaskInfo
+
+		// find preemptable tasks which appear on none revocable node
+		for _, item := range preemptees {
+			if _, ok := item.Pod.Annotations[PreemptableTaskKey]; !ok {
+				continue
+			}
+
+			node, ok := ssn.Nodes[item.NodeName]
+			if !ok {
+				continue
+			}
+
+			if _, ok := node.Node.Labels[RevocableZoneKey]; ok {
+				continue
+			}
+
+			victims = append(victims, item)
+		}
+
+		klog.V(4).Infof("TDM victims are %+v", victims)
+
+		return victims
+	}
+
+	victimsFn := func() []*api.TaskInfo {
```

Comment:
we need support expel step by step, so in victimsFn not return all the victim `taskInfo` in one scheduler period.

Reply:
You are right.